android/libraries/rib-base/src/main/java/com/uber/rib/core/Node.kt

```diff
@@ -0,0 +1,313 @@
+        private set
+    protected var parentViewGroup: ViewGroup? = null
+
+    private var savedViewState: SparseArray<Parcelable> = SparseArray()
+
+
+    private fun generateRibId(): Int =
+        requestCodeRegistry.generateGroupId(tag)
+
+    private fun generateRequestCode(code: Int): Int =
+        requestCodeRegistry.generateRequestCode(tag, code)
+
+    private fun updateRibId(value: Int) {
+        ribId = value
+    }
+
+    fun attachToView(parentViewGroup: ViewGroup) {
+        this.parentViewGroup = parentViewGroup
+        view = createView(parentViewGroup)
+        view?.let {
+            parentViewGroup.addView(it.androidView)
+            it.androidView.restoreHierarchyState(savedViewState)
+            interactor.onViewCreated(it)
+        }
+
+        children.forEach {
+            attachChildView(it)
+        }
+    }
+
+    private fun createView(parentViewGroup: ViewGroup): V? =
+        viewFactory?.invoke(parentViewGroup)
+
+    internal fun attachChild(child: Node<*>, bundle: Bundle? = null) {
+        attachChildView(child)
+        // todo refactor so that this branching is not necessary
+        if (bundle != null) {
+            attachChildNode(child, bundle)
+        } else {
+            attachChildNode(child)
+        }
+    }
+
+    internal fun attachChildView(child: Node<*>) {
+        // parentViewGroup is guaranteed to be non-null if and only if Android view system is available
+        parentViewGroup?.let {
+            child.attachToView(
+                router.getParentViewForChild(child.forClass, view) ?: it
+            )
+        }
+    }
+
+    internal fun detachChild(child: Node<*>) {
+        detachChildNode(child)
+        detachChildView(child)
+    }
+
+    internal fun saveViewState() {
+        view?.let {
+            it.androidView.saveHierarchyState(savedViewState)
+        }
+    }
+
+    internal fun detachChildView(child: Node<*>) {
+        parentViewGroup?.let {
+            child.onDetachFromView(
+                parentViewGroup = router.getParentViewForChild(child.forClass, view) ?: it
+            )
+        }
+    }
+
+    fun onDetachFromView(parentViewGroup: ViewGroup) {
+        children.forEach {
+            detachChildView(
+                child = it
+            )
+        }
+
+        view?.let {
+            parentViewGroup.removeView(it.androidView)
+            interactor.onViewDestroyed()
+        }
+
+        view = null
+        this.parentViewGroup = null
+    }
+
+    /**
+     * Dispatch back press to the associated interactor.
+     *
+     * @return TRUE if the interactor handled the back press and no further action is necessary.
+     */
+    @CallSuper
+    open fun handleBackPress(): Boolean {
+        ribRefWatcher.logBreadcrumb("BACKPRESS", null, null)
+        return router.popBackStack() || interactor.handleBackPress()
+    }
+
+    /**
+     * Called when a node is being attached. Node subclasses can perform setup here for anything
```

Comment:
maybe we should clarify that it happens before router and interactor receive attach event?
Otherwise not clear is it happens before or after.